### PR TITLE
Add debug logging messages on retry

### LINKED ISSUE #514: Emit debug logs on retry?
In my use of AWS.jl, I would find it interesting to know what kinds of network issues we are hitting and how frequently they occur. We could support this by emitting debug logs (e.g. `@debug "AWS.jl retry" reason=...` or something). This wouldn't be spammy since the default console logger drops them, but it would allow setting a custom logger or changing the global log-level to surface them.

### PR DESCRIPTION
**Pull request by ericphanson**
Closes #514

Based on #548

The logging messages are structured so the string is a human-readable summary of the message, and then the fields are  more structured for later analysis, e.g. `retry::Bool`, `reason::String`, etc. Additionally, the id of the log message will be useful for aggregation.

### CODE REVIEW

**Comment by ericphanson** [comment]
Let's see if CI likes it:

bors try

**Comment by ericphanson** [comment]
ok great! I think we should probably first review/merge #548, then merge this on top.

File: test/issues.jl [hunk 1]
```diff
@@ -2,6 +2,15 @@
 
 BUCKET_NAME = "aws-jl-test-issues---" * _now_formatted()
 
+# We try to be insensitive to the existence of other logging messages, so we test
+# by counting the number of messages that meet some criteria
+function log_is_retry(successful)
+    return record ->
+        record.level == Logging.Debug &&
+            haskey(record.kwargs, :retry) &&
+            record.kwargs[:retry] == successful
```
Comment (omus) [inline test/issues.jl:11]:
```suggestion
        record.level == Logging.Debug &&
        haskey(record.kwargs, :retry) &&
        record.kwargs[:retry] == successful
```

File: test/issues.jl [hunk 2]
```diff
@@ -2,6 +2,15 @@
...
+end
```
Comment (omus) [inline test/issues.jl:12]:
```suggestion
function log_is_retry(successful::Bool)
    return record ->
        record.level == Logging.Debug &&
        get(record.kwargs, :retry, nothing) == successful
end
```

**Review by mattBrzezinski — APPROVED** [review]
bors r+

**Review by mattBrzezinski — APPROVED** [review]
bors r+

**Comment by mattBrzezinski** [comment]
bors try

**Comment by mattBrzezinski** [comment]
bors try

**Comment by mattBrzezinski** [comment]
bors try

**Comment by mattBrzezinski** [comment]
Discussed offline, but we'll be revising how to properly introduce debugging here, closing this PR for now.

**Pull request closed without merging**